[GStreamer] Plumb audio sink ID support in MediaPlayer

aperezdc opened the pull request:
#### e1b943ce81dccf779ae6fa4c848d9e1d718b2d7c
<pre>
[GStreamer] Plumb audio sink ID support in MediaPlayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=216880">https://bugs.webkit.org/show_bug.cgi?id=216880</a>

Reviewed by Philippe Normand.

Plumb speaker selection into the GStreamer media playback, both to allow
output/speaker device enumeration, and to choose audio stream destinations
per media element via HTMLMEdiaElement.setSinkId().

Based on a patch initially by Philippe Normand.

* Source/WTF/wtf/glib/GSpanExtras.h:
(WTF::adoptGMallocString): Add helper to adopt heap-allocated C strings
into a GMallocSpan helper.
* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
(WebCore::UserMediaController::voiceActivityDetected): Add missing
ENABLE(MEDIA_SESSION) guard to fix the build when MediaSession is
disabled but MediaStream is enabled.
* Source/WebCore/platform/mediastream/CaptureDevice.h: Make DeviceType a
bit flags enum, to allow usage with OptionSet.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::applyAudioSinkDevice): Added. Helper function to traverse the
audio sink bin recursively to find the actual sinks and change the
device they use as their output.
(WebCore::MediaPlayerPrivateGStreamer::audioOutputDeviceChanged): Added.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
(WebCore::GStreamerAudioCaptureSourceFactory::speakerDevices const): Changed
to delegate the responsibility of enumerating speaker devices to
GStreamerCaptureDeviceManager.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::speakerDevices): Added.
(WebCore::GStreamerCaptureDeviceManager::addDevice): Gather speaker
devices and store them separately, to allow returning them from
the ::speakerDevices() function.
(WebCore::GStreamerCaptureDeviceManager::refreshCaptureDevices): Handle
multiple devices types at once, and add a GstDeviceMonitor filter to
obtain audio output devices from the monitor.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
Make GStreamerCaptureDeviceManager::deviceTypes() return an OptionSet.

Canonical link: <a href="https://commits.webkit.org/289783@main">https://commits.webkit.org/289783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a31b4708fff533e064292866dbaf9199aebe39e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7325 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92674 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38559 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89860 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7706 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15498 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67802 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25548 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90811 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5890 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79451 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48170 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5677 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33868 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37666 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80607 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76075 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34738 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94564 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86584 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14977 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11019 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76647 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75307 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75883 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20256 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18691 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7979 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13725 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14993 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20296 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109078 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14737 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26229 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18181 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16519 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
```diff
@@ -46,8 +47,9 @@ class GStreamerCaptureDeviceManager : public CaptureDeviceManager, public Realti
+    const Vector<CaptureDevice>& speakerDevices();
     const Vector<CaptureDevice>& captureDevices() final;
-    virtual CaptureDevice::DeviceType deviceType() = 0;
+    virtual OptionSet<CaptureDevice::DeviceType> deviceTypes() = 0;
```
I can see how `speakerDevices()` and `captureDevices()` aren't `const`, as they trigger `refreshCaptureDevices()`. 

But `deviceTypes()` could probably be `const` as it seems to return a static set

aperezdc replied:
That works, thanks for the suggestion 👌🏼

spenap commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
```diff
@@ -70,28 +72,37 @@ class GStreamerCaptureDeviceManager : public CaptureDeviceManager, public Realti
 public:
     static GStreamerAudioCaptureDeviceManager& singleton();
-    CaptureDevice::DeviceType deviceType() final { return CaptureDevice::DeviceType::Microphone; }
+    OptionSet<CaptureDevice::DeviceType> deviceTypes() final { return { CaptureDevice::DeviceType::Microphone, CaptureDevice::DeviceType::Speaker }; }
```
here

spenap commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
```diff
@@ -70,28 +72,37 @@ class GStreamerCaptureDeviceManager : public CaptureDeviceManager, public Realti
 public:
     static GStreamerVideoCaptureDeviceManager& singleton();
-    CaptureDevice::DeviceType deviceType() final { return CaptureDevice::DeviceType::Camera; }
+    OptionSet<CaptureDevice::DeviceType> deviceTypes() final { return { CaptureDevice::DeviceType::Camera }; }
```
and here

calvaris reviewed (commented):
LGTM but I guess we should fix that ASSERT.

spenap commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
```diff
@@ -46,8 +47,9 @@ class GStreamerCaptureDeviceManager : public CaptureDeviceManager, public Realti
+    const Vector<CaptureDevice>& speakerDevices();
     const Vector<CaptureDevice>& captureDevices() final;
-    virtual CaptureDevice::DeviceType deviceType() = 0;
+    virtual const OptionSet<CaptureDevice::DeviceType> deviceTypes() const = 0;
```
```suggestion
    virtual OptionSet<CaptureDevice::DeviceType> deviceTypes() const = 0;
```

it's better not to return `const T`: see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-return-const

spenap commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
```diff
@@ -70,28 +72,37 @@ class GStreamerCaptureDeviceManager : public CaptureDeviceManager, public Realti
 public:
     static GStreamerAudioCaptureDeviceManager& singleton();
-    CaptureDevice::DeviceType deviceType() final { return CaptureDevice::DeviceType::Microphone; }
+    const OptionSet<CaptureDevice::DeviceType> deviceTypes() const final { return { CaptureDevice::DeviceType::Microphone, CaptureDevice::DeviceType::Speaker }; }
```
```suggestion
    OptionSet<CaptureDevice::DeviceType> deviceTypes() const final { return { CaptureDevice::DeviceType::Microphone, CaptureDevice::DeviceType::Speaker }; }
```

ditto

spenap commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
```diff
@@ -70,28 +72,37 @@ class GStreamerCaptureDeviceManager : public CaptureDeviceManager, public Realti
 public:
     static GStreamerVideoCaptureDeviceManager& singleton();
-    CaptureDevice::DeviceType deviceType() final { return CaptureDevice::DeviceType::Camera; }
+    const OptionSet<CaptureDevice::DeviceType> deviceTypes() const final { return { CaptureDevice::DeviceType::Camera }; }
```
```suggestion
    OptionSet<CaptureDevice::DeviceType> deviceTypes() const final { return { CaptureDevice::DeviceType::Camera }; }
```

ditto

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
```diff
@@ -246,24 +268,18 @@ void GStreamerCaptureDeviceManager::refreshCaptureDevices()
-            break;
+        auto types = deviceTypes();
+        if (types.contains(CaptureDevice::DeviceType::Camera)) {
+            auto caps = adoptGRef(gst_caps_new_empty_simple("video/x-raw"));
```
Why are we just restricting to raw video cameras? There are not so many that can accept raw video with higher resolutions.

philn replied:
Yes, encoded formats should  also work, the capturer currently decodes those, which isn't great either but is what it is.

aperezdc replied:
This was in the original patch by @philn on which this PR is based, so I assumed it was correct and kept it 😉

philn replied:
Hehe :) Maybe that initial patch was done before decodebin support was added to the capturer. Can you revert this part?

aperezdc replied:
Of course, will do 👌🏼

aperezdc replied:
Done, the EWS is still happy 😄

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -4362,6 +4362,29 @@ void MediaPlayerPrivateGStreamer::checkPlayingConsistency()
+{
+    for (auto* element : GstIteratorAdaptor<GstElement>(GUniquePtr<GstIterator>(gst_bin_iterate_sinks(GST_BIN_CAST(audioSinkBin))))) {
+        // pulsesink and alsasink have a "device" property, whilst pipewiresink has "target-object"
+        if (GST_IS_AUDIO_BASE_SINK(element) && gstObjectHasProperty(element, "device"))
```
This is a bit fragile, it might make sense to check the element factory name instead. But it's ok, we use this `gstObjectHasProperty()` approach elsewhere, it can be improved as a follow-up.

aperezdc replied:
Indeed I tried to do what I could see in other parts of the code, but indeed the check was looking a bit too lax and that's why I added the `GST_IS_*BASE_SINK()` condition as well.

You have a good point, though, so I have filed https://bugs.webkit.org/show_bug.cgi?id=287017 to improve the checks later on.

philn reviewed (approved):
We'll need to check how this works on sandboxed WebProcess when the time comes to integrate Camera portal support :/

aperezdc committed:
[GStreamer] Plumb audio sink ID support in MediaPlayer
https://bugs.webkit.org/show_bug.cgi?id=216880

Reviewed by Philippe Normand.

Plumb speaker selection into the GStreamer media playback, both to allow
output/speaker device enumeration, and to choose audio stream destinations
per media element via HTMLMEdiaElement.setSinkId().

Based on a patch initially by Philippe Normand.

* Source/WTF/wtf/glib/GSpanExtras.h:
(WTF::adoptGMallocString): Add helper to adopt heap-allocated C strings
into a GMallocSpan helper.
* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
(WebCore::UserMediaController::voiceActivityDetected): Add missing
ENABLE(MEDIA_SESSION) guard to fix the build when MediaSession is
disabled but MediaStream is enabled.
* Source/WebCore/platform/mediastream/CaptureDevice.h: Make DeviceType a
bit flags enum, to allow usage with OptionSet.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::applyAudioSinkDevice): Added. Helper function to traverse the
audio sink bin recursively to find the actual sinks and change the
device they use as their output.
(WebCore::MediaPlayerPrivateGStreamer::audioOutputDeviceChanged): Added.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
(WebCore::GStreamerAudioCaptureSourceFactory::speakerDevices const): Changed
to delegate the responsibility of enumerating speaker devices to
GStreamerCaptureDeviceManager.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::speakerDevices): Added.
(WebCore::GStreamerCaptureDeviceManager::addDevice): Gather speaker
devices and store them separately, to allow returning them from
the ::speakerDevices() function.
(WebCore::GStreamerCaptureDeviceManager::refreshCaptureDevices): Handle
multiple devices types at once, and add a GstDeviceMonitor filter to
obtain audio output devices from the monitor.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
Make GStreamerCaptureDeviceManager::deviceTypes() return an OptionSet.

Canonical link: https://commits.webkit.org/289783@main
